### DISCUSSION TITLE
new development tasks

### DISCUSSION THREAD

**Discussion by wenboown** [SeineAI/SeineSailor#23] [Ideas]
1. review.py
Document the logic

2. issue, discussion events
we need to develop new features: what tasks? (Summarize the discussions, brainstorm ideas, code gen based on instructions.)
We need a LLM Agent approach.

3. code review
not just summary, but also unit test.
The PR request needs to trigger the automatic generation of unit testing and e2e testing code in the same PR if the PR does not include necessary testing code.

4. The Code Review includes naive code review includes:

Code redundancy checking
Code naming and convention
Code restructuring, etc.
The code suggestions model includes code changes suggested to the above issues.

5. Integrate Tree-sitter package:
bot action or github action?
store somewhere in the repo (maybe .seine/code-map?)

6. Test the different events

7. check the concurrency_limits

**Comment by wenboown** [comment]
samyuanibm (May 17, 2024, 8:51 PM)
ok
API requests - You can execute up to 1,000 requests to the GitHub API in an hour across all actions within a repository. If requests are exceeded, additional API calls will fail which might cause jobs to fail.
https://docs.github.com/en/actions/learn-github-actions/usage-limits-billing-and-administration just list information here as reminder.
Hope the file download will not … touch the 1k request limitation for GitHub...
 
Huamin Chen (May 17, 2024, 8:56 PM)
In the future, let’s use web hook, this will be more scalable https://docs.github.com/en/webhooks/about-webhooks
 
Huamin Chen (May 17, 2024, 9:01 PM)
https://github.com/SeineAI/SeineSailor/tree/main/tests

**Comment by wangchen615** [comment]
LLM Agent framework: [dspy](https://github.com/stanfordnlp/dspy)
LLM Function call framework: [gorilla](https://github.com/ShishirPatil/gorilla)
Structured Generation Language for LLM: [sglang](https://github.com/sgl-project/sglang)

**Reply by wenboown** [reply]
I took a look at sglang. Its frontend is similar idea as Microsoft's [guidance](https://github.com/guidance-ai/guidance) (Guidance's underlaying mechanism is to access logits of LLM server to control the output or "select".) 
But they all requires the backend inference engine supports (providing them access to the information they need for constrained decoding). I asked the BAM/Watsonx team about supporting `guidance` many months ago, but the answer is no... 

The sglang only supports their own engine and a few other cloud APIs (OpenAI, Anthropic and VertexAI (Gemini)). Watsonx is not supported...
